projects/js-packages/connection/components/disconnect-dialog/index.jsx

```diff
@@ -45,54 +69,211 @@ const DisconnectDialog = props => {
+	 * Should run when we have a connected user.
 	 */
-	const openModal = useCallback(
-		e => {
-			e && e.preventDefault();
-			setOpen( true );
-		},
-		[ setOpen ]
-	);
+	useEffect( () => {
+		if ( connectedUser && connectedUser.ID && connectedUser.login ) {
+			jetpackAnalytics.initialize( connectedUser.ID, connectedUser.login );
+		}
+	}, [ connectedUser, connectedUser.ID, connectedUser.login ] );
 
 	/**
-	 * Close the Disconnect Dialog.
+	 * Run when the disconnect dialog is opened
 	 */
-	const closeModal = useCallback(
-		e => {
-			e && e.preventDefault();
-			setOpen( false );
+	useEffect( () => {
+		if ( isOpen ) {
+			jetpackAnalytics.tracks.recordEvent( 'jetpack_disconnect_dialog_open', defaultTracksArgs );
+		}
+	}, [ isOpen, defaultTracksArgs ] );
+
+	/**
+	 * Keep track of the steps that are presented
+	 */
+	useEffect( () => {
+		if ( isDisconnected && ! isProvidingFeedback && ! isFeedbackProvided ) {
+			jetpackAnalytics.tracks.recordEvent(
+				'jetpack_disconnect_dialog_disconnect_confirm_step',
+				defaultTracksArgs
+			);
+		} else if ( isProvidingFeedback && ! isFeedbackProvided ) {
+			jetpackAnalytics.tracks.recordEvent(
+				'jetpack_disconnect_dialog_survey_step',
+				defaultTracksArgs
+			);
+		} else if ( isFeedbackProvided ) {
+			jetpackAnalytics.tracks.recordEvent(
+				'jetpack_disconnect_dialog_thank_you_step',
+				defaultTracksArgs
+			);
+		}
+	}, [ isDisconnected, isProvidingFeedback, isFeedbackProvided, defaultTracksArgs ] );
+
+	/**
+	 * Disconnect the site.
+	 * Uses the rest API to remove the Jetpack connection.
+	 */
+	const _disconnect = useCallback( () => {
+		restApi
+			.disconnectSite()
+			.then( () => {
+				setIsDisconnecting( false );
+				setIsDisconnected( true );
+			} )
+			.catch( error => {
+				setIsDisconnecting( false );
+				setDisconnectError( error );
+
+				if ( onError ) {
+					onError( error );
+				}
+			} );
+	}, [ setIsDisconnecting, setIsDisconnected, setDisconnectError, onError ] );
+
+	/**
+	 * Submit the optional survey following disconnection.
+	 */
+	const _submitSurvey = useCallback(
+		( surveyData, tracksSurveyData ) => {
+			// Send survey response to wpcom
+			const base = 'https://public-api.wordpress.com';
+			const path = '/wpcom/v2/marketing/feedback-survey';
+			const method = 'POST';
+
+			setIsSubmittingFeedback( true );
+
+			// We cannot use `@wordpress/api-fetch` here since it unconditionally sends
+			// the `X-WP-Nonce` header, which is disallowed by WordPress.com.
+			// If the submission receives an error, there's not really anything the user is able to do to fix it.
+			// In these cases, just go ahead and show the last survey step.
+			fetch( base + path, {
+				method: method,
+				headers: {
+					'Content-Type': 'application/json',
+					Accept: 'application/json',
+				},
+				body: JSON.stringify( surveyData ),
+			} )
+				.then(
+					result => {
+						return result.json();
+					},
+					// Error in sending the data.
+					() => {
```

Comment:
Maybe a naive question, but if we hit an error during the request and execute this reject function, will chaining the next `then()` call cause an error because this isn't returning a promise or other thenable object?

Reply:
Great question. I'm still learning about the intricacies of Promises myself...
I think that if the reject function returns nothing, then the subsequent `then` will receive `undefined` as its value. Which would not be desirable and cause some errors when checking properties on `undefined`.
I'm updating this promise chain with a `.catch` to handle any errors in the chain.